pages/en/lb4/Testing-Your-Extensions.md

```diff
@@ -0,0 +1,303 @@
+    };
+  }
+}
+```
+
+**`test/unit/providers/random-number.unit.test.ts`**
+```ts
+import {RandomNumberProvider} from '../../..';
+import {expect} from '@loopback/testlab';
+
+describe('RandomNumberProvider (unit)', () => {
+  it('generates a random number within range', () => {
+    const provider = new RandomNumberProvider().value();
+    const random: number = provider(3);
+
+    expect(random).to.be.a.Number();
+    expect(random).to.equalOneOf([1, 2, 3]);
+  })
+})
+```
+
+#### Repositories
+
+*TBD ... How / What?*
+
+### Integration Tests
+
+An integration test plays an important part in your test suite by ensuring your
+extension artifacts play well with each other as well as `@loopback`. It is
+recommended to test two items together and other integrations can be
+substituted by test doubles so it becomes apparent where the integration breaks
+(if it does).
+
+#### Controllers
+
+*TBD ... Repositories? Other Services?*
+
+#### Decorators
+
+*TBD ... How / What?*
+
+#### Mixins
+
+As mentioned under unit/mixins section, a Mixin extends a base Class by returning
+an anonymous class. As such to test a Mixin, it should be done by actually
+using the mixin with it's base Class. Since this requires two Classes to work
+together, an integration test will be needed. A Mixin should be tested to ensure
+new / overridden methods exist and work as expected in the new Mixed class. 
+
+**`src/mixins/time.mixin.ts`**
+```ts
+import {Constructor} from '@loopback/context';
+export function TimeMixin<T extends Constructor<any>>(superClass: T) {
+  return class extends superClass {
+    constructor(...args: any[]) {
+      super(...args);
+      if (!this.options) this.options = {};
+
+      if (typeof this.options.timeAsString !== 'boolean') {
+        this.options.timeAsString = false;
+      }
+    }
+
+    time() {
+      if (this.options.timeAsString) {
+        return new Date().toString();
+      }
+      return new Date();
+    }
+  };
+}
+```
+
+**`test/integration/mixins/time.intg.test.ts`**
+```ts
+import {expect} from '@loopback/testlab';
+import {Application} from '@loopback/core';
+import {TimeMixin} from '../../..';
+
+describe('TimeMixin (integration)', () => {
+  it('mixed class has .time()', () => {
+    const myApp = new AppWithTime();
+    expect(typeof myApp.time).to.be.eql('function');
+  });
+
+  it('returns time as string', () => {
+    const myApp = new AppWithLogLevel({
+      timeAsString: true,
+    });
+
+    const time = myApp.time();
+    expect(time).to.be.a.String();
+  });
+
+  it('returns time as Date', () => {
+    const myApp = new AppWithLogLevel();
+
+    const time = myApp.time();
+    expect(time).to.be.a.Date();
+  });
+
+  class AppWithTime extends TimeMixin(Application) {}
+});
+```
+
+#### Providers
+
+*TBD ... How / What?*
+
+#### Repositories
+
+*TBD ... How / What*
+
+### Acceptance Test
+
+An Acceptance test is the ultimate test. This is a comprehensive test, written
```

Comment:
`This is a comprehensive test, written end-to-end that is responsible ...`
Grammar: This is a comprehensive test written end-to-end **which** is reponsible ...
The comma needs to be either moved after 'end-to-end' or gotten rid of entirely (I'm not sure)